File: c++/language_constructs.cpp
#/* run as shell script: sh $0
set -x; c++ -o ${0%%.*} $0; exit
*/
//
// Copyright (c) 2019 B.Roden
//
// MIT License
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

// Scope: basic program structure, control structures, functions, arrays, pointers, file I/O, processs/threads, and basic concepts of object-oriented programming (classes, objects, function overloading) 

// Importing/Including
#include <iostream>
using namespace std;
#include <unistd.h>
#include <signal.h>
#include <sys/wait.h>

// Definitions
#define ASCII_BASE '0'

// Declarations - Global Variable
int i=-1;

// Declarations - Class
class Rectangulum {
    union{ int width, x; };
    union{ int height, y; };
    union{ int depth, z; };
public:
    int set(int,int); int set(int,int,int);
    int get(int*,int*); int get(int*,int*,int*);
    int area(){return width*height;}
    int volume(){return width*height*depth;}
};
int Rectangulum::set(int x, int y) {
    try { width = x; height = y; } catch (int e) { return e; }
    return 0;
}
int Rectangulum::set(int x, int y, int z) {
    try { width = x; height = y; depth = z; } catch (int e) { return e; }
    return 0;
}
int Rectangulum::get(int *x, int *y) {
    try { *x = width; *y = height; } catch (int e) { return e; }
    return 0;
}
int Rectangulum::get(int *x, int *y, int *z) {
    try { *x = width; *y = height; *z = depth; } catch (int e) { return e; }
return 0;
}

// Functions
void logrecord(const char *format, ...)
{
    va_list args;
    va_start(args, format);
    vprintf(format, args);
    va_end(args);
}

void cbTerminate(int i){logrecord("signal [%d %s]\n",i,(i==15)?"SIGTERM":"");exit(i);}

// Main
int main(int argc, char *argv[]) {
    cout << "***" << "START\n";

// Parameters
    if (argc>1) {
        cout << "argv[1] == \"" << argv[0] << "\" argv[1] == " << argv[1] << endl;
        i = (int)*argv[argc-1] - ASCII_BASE;                // only first char/byte of string
		i = static_cast<int>(*argv[argc-1]) - ASCII_BASE;   // only first char/byte of string
    i = atoi(argv[argc-1]);
    }

// External
    system("uname -a");

// Object - Data & Methods
    Rectangulum r;
    int x,y,z;

    r.set(3,4);
    r.get(&x,&y)?cout <<"r.get rc != 0\n":cout <<"r.get rc == 0\n";
    printf("values: x == %d y == %d\n",x,y);
    cout << "area: " << r.area() << endl;

    r.set(3,4,2);
    r.get(&x,&y,&z)?cout <<"r.get rc != 0\n":cout <<"r.get rc == 0\n";
    printf("values: x == %d y == %d z == %d\n",x,y,z);
    cout << "volume: " << r.volume() << endl;

// Control - Conditionals
    if (i>0) printf("1: i is %d>0\n",i);
    else if (i<0) printf("1: i is %d<0\n",i);
    else printf("1: i is %d==0\n",i);

    if (i>0||i<0) printf("2: i is %d!=0\n",i);

    (i!=0)?printf("3: i is %d!=0\n",i):printf("3: i is %d==0\n",i);

    switch(i) {
    case 1 ... 9999:
        printf("4: i is %d>0\n",i); break;
    case 0:
        printf("4: i is %d==0\n",i); break;
    default:
        printf("4: i is %d<0\n",i); break;
    }

// Control - Loops - Sequential
    const int k=3;

    for (int j=0;j<=k;j++) printf("for k==%d j==%d\n",k,j);

    int j=0;
    while (j<=k) printf("while k==%d j==%d\n",k,j++);

    j=0;
    do printf("do while k==%d j==%d\n",k,j); while (++j<=k);

// Control - Loops - Breakout variations
    j=0;
    do {
        if (j>k) break;
        printf("while w/break j==%d\n",j);
    } while (++j);

    j=0;
    do {
        if (j>k) {j=-1;continue;};
        printf("while w/continue j==%d\n",j);
    } while (++j);

    j=0;
    do {
        if (j>k) goto end_loop;
        printf("while w/goto j==%d\n",j);
    } while (++j);
    end_loop:

    try {
        j=0;
        do {
            if (j>k) throw 0;
             printf("while w/throw j==%d\n",j);
        } while (++j);
    } catch (int e){}

// Control - Thread/Process
    {
        pid_t  p1,p2; int rc = 0;
        if ((p1 = fork())<0) {
            printf("Fork failed %d\n",p1);
        } else if (p1 == 0){
            printf("In child process %d\n",p1);
        } else {
            printf("In parent process with child pid = %d\n",p1);
            p2 = wait(&rc);
            (WIFEXITED(rc))?printf("Child process pid %d exit status = %d\n",p2,WEXITSTATUS(rc)):NULL;
            exit(0);
        }
    }

// Control - Exceptions
    signal (SIGTERM, cbTerminate);

    int _e_;
    try {
        throw i;
    } catch (int e) {
        cout << "Exception e == " << e << "\n";
        _e_ = e;
        if (e!=9) goto terminate;
    }

    cout << "Bye\n"; exit(0);

    terminate:
        switch(_e_) {
            case 1: cout << "Terminate case 1\n"; break;
            default: cout << "Terminate default\n"; raise(SIGTERM);
        }
}

// I/O (File & Network)

File: python/language_constructs.py
#
# Copyright (c) 2019 B.Roden
#
# MIT License
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#

# Scope: basic program structure, control structures, functions, arrays, pointers, file I/O, processs/threads, and basic concepts of object-oriented programming (classes, objects, function overloading)
# https://docs.python.org/3.7/library/index.html

# Importing/Including
import os,sys,signal,threading,getopt
import http.client
import ftplib

try: from platform import system as platform
except: sys.exit("***ENOIMPORT" % "platform")

# Definitions

# Global Variables
i:int

# Classes

# Functions
def logrecord(format,*argv):
 print(format % argv);

def cbTerminate(i,_):
 if i == 15: s = "SIGTERM"
 logrecord("signal [%d %s]",i,s)
 sys.exit(i)


# Main
if __name__ == "__main__":
 print("***START")

# Parameters
 if len(sys.argv) > 1:
  for i in range(len(sys.argv)): print("arg[%d]=%s" % (i,sys.argv[i]))
  i = int(sys.argv[1])

# External
 os.system('''uname -a''')

# Data Structures

# Conditionals
 if (i>0):   print("1: i is %d>0" % i)
 elif (i<0): print("1: i is %d<0" % i)
 else:       print("1: i is %d==0" % i)

 if (i>0 or i<0): print("2: i is %d!=0" % i)

 i_is_not_zero_rule = [i>0, i<0]
 if (not all(i_is_not_zero_rule)): print("2b: i is %d!=0" % i)

 print("3: i is %d!=0" % i) if (i != 0) else print("3: i is %d==0" % i)

 if (i > 0 and i < 10000):
  print("4: i is %d>0" % i)
 elif (i == 1):
  print("4: i is %d==0" % i)
 else:
  print("4: i is %d<0" % i)

# Loops - Sequential
 k=3
 for j in range(0,k+1): print("for k==%d j==%d" % (k,j))

 j=0
 while (j<=k):
  print("while k==%d j==%d" % (k,j))
  j+=1

## N/A do..until
 j=0
 while True:
  print("do while k==%d j==%d" % (k,j))
  j+=1
  if (j > k): break

# Loops - Breakout variations

 j=0
 while True:
  if (j > k): break
  print("while w/break k==%d j==%d" % (k,j))
  j+=1

 j=0
 while j >= 0:
  if (j>k):
   j=-1
   continue
  print("while w/continue k==%d j==%d" % (k,j))
  j+=1

## N/A do..until

 try:
  j=0
  while True:
   print("while w/throw k==%d j==%d" % (k,j))
   j+=1
   if (j>k): raise 0
 except: pass

# Control - Thread/Process
 p1 = os.fork()
 if p1 < 0:
  print("Fork failed %d"%p1)
 elif p1 == 0:
  print("In child process %d"%p1)
 else:
  print("In parent process child pid = %d"%p1);
  p2,rc = os.wait()
  print("Child process pid %d exit status = %d"%(p2,rc))
  sys.exit(0)

# Exceptions
 signal.signal(signal.SIGTERM, cbTerminate)

 try: 0/i
 except:
  print("Terminate 0")
  signal.pthread_kill(threading.get_ident(),signal.SIGTERM)
 else:
  print("Terminate %d" % i)

# I/O (File)
 try:
  with open(sys.argv[2],"r+") as fd:
   try: file_data = fd.read()
   except: sys.exit('***EREAD')
   else:
    print("Read \"%s\" from %s"%(file_data,sys.argv[2]))
    try: rc = fd.write("hello")
    except: sys.exit('***EWRITE ' + sys.argv[2] + rc)
    else: print("Wrote %d byte to %s"%(rc,sys.argv[2]))
   try: rc = fd.close()
   except: sys.exit('***ECLOSE ' + sys.argv[2] + rc)
 except: sys.exit('***EOPEN ' + sys.argv[2])

# I/O (TCP/HTTP)
 cd = http.client.HTTPSConnection("www.google.com")
 cd.request("GET", "/")
 rc = cd.getresponse()
 print(rc.status, rc.reason)
 html_data = rc.read()
 print("Read http data = \"%80.80s ...\""%html_data)

 try: ftp = ftplib.FTP(host="ftp1.at.proftpd.org",timeout=3)
 except: sys.exit('***EFTP')
 else:
  try: print(ftp.login())
  except: sys.exit('***EFTPLOGIN')
  else:
   try: ftp.retrlines('LIST')
   except: sys.exit('***EFTPLIST')
   else:
    print(ftp.cwd('distrib'))
    try: ftp.retrlines('LIST')
    except: sys.exit('***EFTPLIST')


